src/include/config-h.in.cmake

```diff
@@ -87,6 +87,12 @@
 /* Define if you have fuse */
 #cmakedefine HAVE_LIBFUSE
 
+/* Define version major */
+#define CEPH_FUSE_VERSION_MAJOR @FUSE_VERSION_MAJOR@
```

Comment:
no, this does not work. the funny thing about cmake substitution is that the macro name should be consistent with that of its value. is it possible to have something like
```c++
#ifdef FUSE_VERSION_MINOR
#undef FUSE_VERSION_MINOR
#endif
#define FUSE_VERSION_MINOR @FUSE_VERSION_MINOR@
```

Reply:
the #undef here works, but with this we must make sure that all the `X.cc` files who are including the `fuse_common.h` header must be in front of `acconfig.h` header file:

```
#include <fuse_common.h>
#include "acconfig.h"
```

I was just thinking with a new name as `CEPH_FUSE_VERSION_MINOR` then we could get rid of  the header order issue.

Will it make sense ?

Reply:
actually this ordering 

> 7. Other libraries' .h files.
> 8. Your project's .h files.

is suggested by https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes . 

but yes, i think both approaches have their pros and cons respectively:

- CEPH_FUSE_VERSION_MINOR: more explicit this way. and probably simpler.
- FUSE_VERSION_MINOR: minimal changes. so we can limit the scope of the fix of wrong FUSE_VERSION in `fuse_common.h` in acconfig.h, and we don't have to s/FUSE_VERSION/CEPH_FUSE_VERSION/ for addressing this bug.

so i don't feel strong either way.

Reply:
Okay, let me fix it and add some comments about this.

Reply:
I just undef the FUSE_MINOR_VERSION in src/include/ceph_fuse.h:

```
 #include <fuse.h>

#ifdef FUSE_MINOR_VERSION
#undef FUSE_MINOR_VERSION
#endif

#include "acconfig.h"
```

Then once the user has included the "ceph_fuse.h" header file, we can get rid of the header order issue, and both the following two cases will work:

```
#include <fuse.h>
#include "include/ceph_fuse.h"
```

or 

```
#include "include/ceph_fuse.h"
#include <fuse.h>
```

Reply:
Sorry, this still have some problems.

In case that:

```
#include "int_types.h"  <==== it seems this one must be ahead of all the others ? And it will include "acconfig.h" header file.

#include <...>
#include "include/ceph_fuse.h"
```
In this case the "acconfig.h" will be always ahead of fuse.h. 

This is making this more complicated.

Reply:
And we are getting "redefined" warnings here.

Reply:
And even we do the undef in "acconfig.h":

```
#ifdef FUSE_VERSION_MINOR
#undef FUSE_VERSION_MINOR
#endif
#define FUSE_VERSION_MINOR @FUSE_VERSION_MINOR@
```

We are still hitting the same issue.

Reply:
Let's do undef the FUSE_VERSION instead, then it will simple:

```
#include <fuse.h>
#include "acconfig.h"

#ifdef FUSE_VERSION
#undef FUSE_VERSION
#define FUSE_VERSION FUSE_MAKE_VERSION(CEPH_FUSE_MAJOR_VERSOIN, CEPH_FUSE_MINOR_VERSOIN)
#endif

```

With this we can also get rid of the file header order issue.